tensorflow/python/framework/op_def_library.py

```diff
@@ -618,8 +618,8 @@ def apply_op(self, op_type_name, name=None, **keywords):
         if input_arg.is_ref:
           if not all(x._is_ref_dtype for x in types):  # pylint: disable=protected-access
             raise TypeError(
-                "Input '%s' of '%s' Op requires l-value input" %
-                (input_name, op_type_name))
+                "'%s' Op requires that input '%s' be tf.Variable" %
+                (op_type_name, input_name))
```

Comment:
I think this error message could be misleading. Sure, `tf.Variable` is the most common "l-value input", but you can also pass the result of `tf.assign()` to another `tf.assign()` as the l-value, in order to chain updates.

How about instead making the error message:

"'%s' op requires that input '%s' be a mutable tensor (e.g a tf.Variable)"